File: Python/balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
    
        def maxDepth(root: Optional[TreeNode]) -> int:
            if not root:
                return [True, 0]
            l = maxDepth(root.left)
            r = maxDepth(root.right)
            balanced = l[0] and r[0] and abs(r[1] - l[1]) <= 1
            return [balanced, 1 + max(l[1], r[1])]

        return maxDepth(root)[0]

File: Cpp/balanced_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool trip = true;

    bool isBalanced(TreeNode* root) {
        helper(root);
        return trip;        
    }
    int helper(TreeNode* root){
        if(root == NULL) return 0;
        int l, r;
        l = helper(root -> left);
        r = helper(root -> right);
        if(abs(l-r)>1) trip = false;
        return max(l,r) + 1;
    }
};